sqlglot/generator.py

```diff
@@ -695,25 +704,55 @@ def properties(
         prefix: str = "",
         sep: str = ", ",
         suffix: str = "",
-        before: bool = False,
+        wrapped: bool = True,
     ) -> str:
         if properties.expressions:
             expressions = self.expressions(properties, sep=sep, indent=False)
-            expressions = expressions if before else self.wrap(expressions)
+            expressions = self.wrap(expressions) if wrapped else expressions
             return f"{prefix}{' ' if prefix and prefix != ' ' else ''}{expressions}{suffix}"
         return ""
 
     def with_properties(self, properties: exp.Properties) -> str:
         return self.properties(properties, prefix=self.seg("WITH"))
 
+    def locate_properties(self, properties: exp.Properties) -> dict:
+        properties_locs = {}
```

Comment:
can you do this in one pass in stead of n passes?

for property in properties.expressions:
   if ...
     locations[...] = ..
  elif...